auth/auth.go

```diff
@@ -0,0 +1,281 @@
+package auth
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/coreos/go-oidc"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"golang.org/x/oauth2"
+
+	"github.com/sourcegraph/sourcegraph-accounts-sdk-go/scopes"
+)
+
+// Config contains the configuration for the SAMS authentication handler.
+type Config struct {
+	// Issuer is the SAMS instance URL, e.g. "https://accounts.sourcegraph.com".
+	Issuer string
+	// ClientID is the SAMS client ID, e.g. "sams_cid_xxx".
+	ClientID string
+	// ClientSecret is the SAMS client secret, e.g. "sams_cs_xxx".
+	ClientSecret string
+	// RequestScopes is the list of requested scopes for access tokens that are
+	// issued to this client.
+	RequestScopes []scopes.Scope
+	// RedirectURI is the URL to redirect to after the user has authenticated.
+	RedirectURI string
+
+	// FailureHandler is the HTTP handler to call when an error occurs. Use
+	// ErrorFromContext to extract the error.
+	FailureHandler http.Handler
+	// StateSetter sets the randomly-generated state to the per-user session.
+	StateSetter func(w http.ResponseWriter, r *http.Request, state string) error
+	// StateGetter gets the state from the per-user session.
+	StateGetter func(r *http.Request) (string, error)
+	// StateDeleter deletes the state from the per-user session.
+	StateDeleter func(w http.ResponseWriter, r *http.Request)
```

Comment:
⚠️  Something seems a little off here. Why do `StateSetter` and `StateDeleter` require the `http.ResponseWriter`? It seems like neither of these should "own" writing the HTTP response. (e.g. if either of them writes the HTTP status code, then any attempts to write HTTP headers later will panic.)

I'm guessing that it is required because otherwise it wouldn't be possible to delete any cookies that the user presented in the HTTP request? But if so, isn't that explicitly not something you want to encourage? So maybe you should just intentionally not make it possible? And only have the `StateDeleter` and `StateSetter` take the `http.Request`?

Reply:
See https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/8#issuecomment-2030383694 

Should have addressed this comment.